### PR TITLE
Dave logic

### PR DESCRIPTION
**Pull request by sshang0309**

### CODE REVIEW

**Comment by sshang0309** [comment]
You told me that I need to add a HashSet for the commandSet "undo" and "exit", but I have no idea what they are used for, so I didn't add it. How can I incorporate this kind of thing?

File: src/lifetracker/logic/LogicImpl.java [hunk 1]
```diff
@@ -8,27 +7,17 @@
...
     private static final String ERROR_SAVE = "Warning: There was an error saving to the save file!";
     private static final String ERROR_INVALID_COMMAND = "Error: Command was not a valid command!";
 
-    private static final FormatStyle DATE_STYLE = FormatStyle.MEDIUM;
-    private static final FormatStyle TIME_STYLE = FormatStyle.SHORT;
-
     private Parser commandParser;
     private Storage calendarStorage;
     private CalendarList calendar;
+    private Stack<String> commandStack;
```
Comment (yichenshen) [inline src/lifetracker/logic/LogicImpl.java]: commandStack should store a stack of CommandObjects, not Strings

File: src/lifetracker/logic/LogicImpl.java [hunk 2]
```diff
@@ -49,18 +40,30 @@ public ExecuteResult executeCommand(String commandString) {
...
             return runResult;
+
         } else {
             CommandObject commandToExecute;
             CalendarList executedState;
 
-            try {
+            if (commandString.equals("undo")) {
+
+                commandStack.pop();
```
Comment (yichenshen) [inline src/lifetracker/logic/LogicImpl.java]: Imagine what happens here. The user inputs "add meeting", then "undo". The code here will just execute "add meeting" again, which is wrong.

What should happen is that you store the CommandObject into the stack. When "undo" is called, pop out a CommandObject, then call it's undo function.

**Comment by yichenshen** [comment]
For For HashSet/HashMap, the idea was that you store the commands in a HashMap with method references so you don't have to write one if statement for every single command.

It allows for better SLAP

e.g.

``` Java
Map<String, Supplier<ExecuteResults>> commandMap = new HashMap<>();

commandMap.put("undo", this::undoLastCommand);
commandMap.put("exit", this::exitProgram);

//somewhere below

private ExecuteResult undoLastCommand(){
    //Execute the undo and return
}

private ExecuteResult exitProgram(){
    //Return exit result
}

//When dealing with commandString:
String trimmedCommand = commandString.trim();
ExecuteResults returnResults;

if(commandMap.containsKey(trimmedCommand)){
    returnResults = commandMap.get(trimmedCommand).get();
} else{
   //Normal parser stuff
}
```

You can Google method references, as well as refer to ParserImpl.

This is not necessary at all for the code to work. Just some idea to make it neater and easier to add new commands.

File: src/lifetracker/logic/LogicImpl.java [hunk 3]
```diff
@@ -49,18 +40,28 @@ public ExecuteResult executeCommand(String commandString) {
...
+                try {
+                    commandToExecute = commandParser.parse(commandString);
+                    executedState = commandToExecute.execute(calendar);
+                } catch (IllegalArgumentException ex) {
+                    ExecuteResult errorResult = new CommandLineResult();
+                    errorResult.setComment(ERROR_INVALID_COMMAND);
+                    return errorResult;
+                }
+
+                commandStack.push(commandToExecute);
```
Comment (yichenshen) [inline src/lifetracker/logic/LogicImpl.java:64]: Don't push into stack if the command was invalid. It'll cause an error upon undoing.

**Comment by sshang0309** [comment]
Yes. As shown in the code, if the command is invalid, an exception will be thrown and return the errorResult to the UI. The push step will not be executed.

**Comment by yichenshen** [comment]
Ah ok, sorry about that, missed it.

**Pull request merged**